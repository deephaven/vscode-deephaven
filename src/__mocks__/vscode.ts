/**
 * Mock `vscode` module. Note that `vi.mock('vscode')` has to be explicitly
 * called in any test module needing to use this mock. It will not be loaded
 * automatically.
 */
import { vi } from 'vitest';

export const workspace = {
  getConfiguration: vi.fn().mockReturnValue(new Map()),
};
