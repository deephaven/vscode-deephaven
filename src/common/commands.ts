import { EXTENSION_ID } from './constants';

export const CONNECT_TO_SERVER_CMD = `${EXTENSION_ID}.connectToServer`;
export const CREATE_NEW_TEXT_DOC_CMD = `${EXTENSION_ID}.createNewTextDoc`;
export const DISCONNECT_EDITOR_CMD = `${EXTENSION_ID}.disconnectEditor`;
export const DISCONNECT_FROM_SERVER_CMD = `${EXTENSION_ID}.disconnectFromServer`;
export const DOWNLOAD_LOGS_CMD = `${EXTENSION_ID}.downloadLogs`;
export const OPEN_IN_BROWSER_CMD = `${EXTENSION_ID}.openInBrowser`;
export const REFRESH_SERVER_TREE_CMD = `${EXTENSION_ID}.refreshServerTree`;
export const REFRESH_SERVER_CONNECTION_TREE_CMD = `${EXTENSION_ID}.refreshServerConnectionTree`;
export const RUN_CODE_COMMAND = `${EXTENSION_ID}.runCode`;
export const RUN_SELECTION_COMMAND = `${EXTENSION_ID}.runSelection`;
export const SELECT_CONNECTION_COMMAND = `${EXTENSION_ID}.selectConnection`;
export const START_SERVER_CMD = `${EXTENSION_ID}.startServer`;
export const STOP_SERVER_CMD = `${EXTENSION_ID}.stopServer`;
