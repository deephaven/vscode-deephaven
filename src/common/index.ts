export * from './constants';
export type * from './types';
