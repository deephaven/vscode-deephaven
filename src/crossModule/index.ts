export * from './assertUtil';
export * from './constants';
