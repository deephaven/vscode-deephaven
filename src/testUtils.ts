/**
 * This module contains test utils that should only be consumed by `.spec.ts`
 * files, but since it's not a `.spec.ts` file, it is special cased to exclude
 * from the extension build (tsconfig.json) and include in the unit test build
 * (tsconfig.unit.json). It shouldn't be imported by the extension code or have
 * any dependencies on the extension code.
 */

import { expect, vi } from 'vitest';
import type * as vscode from 'vscode';
export const bitValues = [0, 1] as const;
export const boolValues = [true, false] as const;

export type MessageHandler<TData> = (
  event: Partial<MessageEvent<TData>>
) => unknown;

/**
 * Return last registered event handler for 'message' event. Assumes that
 * `window.addEventListener` has been mocked or spied on.
 */
export function getLastMessageHandler<TData>(): MessageHandler<TData> {
  const messageCalls = vi
    .mocked(window.addEventListener<'message'>)
    .mock.calls.filter(([type]) => type === 'message');

  expect(messageCalls.length).toBeGreaterThan(0);

  return messageCalls.at(-1)![1] as MessageHandler<TData>;
}

/**
 * Generate a 2 dimensional array of all possible combinations of the given value
 * lists.
 *
 * e.g.
 * matrix([1, 2], ['a', 'b']) => [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']]
 *
 * matrix([1, 2], ['a', 'b', 'c']) => [
 *  [1, 'a'], [1, 'b'], [1, 'c'],
 *  [2, 'a'], [2, 'b'], [2, 'c'],
 * ]
 *
 * @param args Value lists
 * @returns 2D array of all possible combinations
 */
export function matrix<
  TArgs extends (unknown[] | readonly unknown[])[],
  TReturn extends { [P in keyof TArgs]: TArgs[P][number] },
>(...args: TArgs): TReturn[] {
  const [first, ...rest] = args;

  if (rest.length === 0) {
    return first.map(value => [value] as TReturn);
  }

  // recursively call matrix
  const restMatrix = matrix(...rest);

  return first.flatMap(value =>
    restMatrix.map(values => [value, ...values] as TReturn)
  );
}

function lineAt(this: vscode.TextDocument, line: number): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  position: vscode.Position
): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  lineOrPosition: number | vscode.Position
): vscode.TextLine {
  const lineNumber =
    typeof lineOrPosition === 'number' ? lineOrPosition : lineOrPosition.line;

  return mockT<vscode.TextLine>({
    lineNumber,
    text: this.getText().split('\n')[lineNumber] ?? '',
  });
}

/** Mock a `vscode.TextDocument` */
export function mockDocument(
  fileName: string,
  version: number = 1,
  content = ''
): vscode.TextDocument {
  return mockT<vscode.TextDocument>({
    fileName,
    uri: mockUri(`file://mock/path/${fileName}`),
    version,
    getText: () => content,
    lineAt,
  });
}

/** Mock an object providing partial properties */
export function mockT<T>(partial: Partial<T>): T {
  return partial as T;
}

/** Mock a `vscode.Uri` */
export function mockUri(path: string): vscode.Uri {
  return mockT<vscode.Uri>({
    path,
    toString: () => path,
  });
}
