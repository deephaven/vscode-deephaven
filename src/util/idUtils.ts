import { customAlphabet, urlAlphabet } from 'nanoid';
import type { UniqueID } from '../types';

// The default nanoid alphabet includes `_`. Using custom alphabet without it
// to make ids group better in `_` delimited strings.
const nanoidCustom = customAlphabet(urlAlphabet.replace('_', ''), 21);

/**
 * Generate a unique id.
 * @param size The size of the id to generate. Defaults to 21 since that is what
 * nanoid uses as its default.
 */
export function uniqueId(size: number = 21): UniqueID {
  return nanoidCustom(size) as UniqueID;
}

/*
 * Create base-64 encoded key from a random string with the length no less than
 * 96 (required by the DH authentication server).
 */
export function makeSAMLSessionKey(): string {
  let key = '';
  for (let i = 0; i < 96; i += 1) {
    key += String.fromCharCode(Math.floor(Math.random() * 255));
  }
  return Buffer.from(key, 'binary').toString('base64');
}
