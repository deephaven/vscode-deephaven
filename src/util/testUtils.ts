import type * as vscode from 'vscode';
export const bitValues = [0, 1] as const;
export const boolValues = [true, false] as const;

/**
 * Generate a 2 dimensional array of all possible combinations of the given value
 * lists.
 *
 * e.g.
 * matrix([1, 2], ['a', 'b']) => [[1, 'a'], [1, 'b'], [2, 'a'], [2, 'b']]
 *
 * matrix([1, 2], ['a', 'b', 'c']) => [
 *  [1, 'a'], [1, 'b'], [1, 'c'],
 *  [2, 'a'], [2, 'b'], [2, 'c'],
 * ]
 *
 * @param args Value lists
 * @returns 2D array of all possible combinations
 */
export function matrix<
  TArgs extends (unknown[] | readonly unknown[])[],
  TReturn extends { [P in keyof TArgs]: TArgs[P][number] },
>(...args: TArgs): TReturn[] {
  const [first, ...rest] = args;

  if (rest.length === 0) {
    return first.map(value => [value] as TReturn);
  }

  // recursively call matrix
  const restMatrix = matrix(...rest);

  return first.flatMap(value =>
    restMatrix.map(values => [value, ...values] as TReturn)
  );
}

function lineAt(this: vscode.TextDocument, line: number): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  position: vscode.Position
): vscode.TextLine;
function lineAt(
  this: vscode.TextDocument,
  lineOrPosition: number | vscode.Position
): vscode.TextLine {
  const lineNumber =
    typeof lineOrPosition === 'number' ? lineOrPosition : lineOrPosition.line;

  return mockT<vscode.TextLine>({
    lineNumber,
    text: this.getText().split('\n')[lineNumber] ?? '',
  });
}

/** Mock a `vscode.TextDocument` */
export function mockDocument(
  fileName: string,
  version: number = 1,
  content = ''
): vscode.TextDocument {
  return mockT<vscode.TextDocument>({
    fileName,
    uri: mockUri(`file://mock/path/${fileName}`),
    version,
    getText: () => content,
    lineAt,
  });
}

/** Mock an object providing partial properties */
export function mockT<T>(partial: Partial<T>): T {
  return partial as T;
}

/** Mock a `vscode.Uri` */
export function mockUri(path: string): vscode.Uri {
  return mockT<vscode.Uri>({
    path,
    toString: () => path,
  });
}
