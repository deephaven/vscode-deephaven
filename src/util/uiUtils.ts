import * as vscode from 'vscode';
import {
  SELECT_CONNECTION_COMMAND,
  STATUS_BAR_CONNECTING_TEXT,
  STATUS_BAR_DISCONNECTED_TEXT,
  ICON_ID,
} from '../common';
import type {
  ConnectionType,
  ConsoleType,
  ConnectionPickItem,
  IDhService,
  ServerState,
  SeparatorPickItem,
} from '../types';

export interface ConnectionOption {
  type: ConnectionType;
  label: string;
  url: URL;
}

export interface DisconnectOption {
  label: string;
  url: null;
}

export interface WorkspaceFolderConfig {
  readonly uri: vscode.Uri;
  readonly name?: string;
}

/**
 * Create quickpick for selecting a connection.
 */
export async function createConnectionQuickPick(
  servers: ServerState[],
  connections: IDhService[],
  editorLanguageId: string,
  editorActiveConnectionUrl?: URL | null
): Promise<
  | ConnectionPickItem<'server', ServerState>
  | ConnectionPickItem<'connection', IDhService>
  | null
> {
  const serverOptions: ConnectionPickItem<'server', ServerState>[] =
    servers.map(data => ({
      type: 'server',
      label: data.url.toString(),
      iconPath: new vscode.ThemeIcon(ICON_ID.server),
      data,
    }));

  const connectionOptions: ConnectionPickItem<'connection', IDhService>[] = [];

  for (const dhService of connections) {
    const isActiveConnection =
      editorActiveConnectionUrl?.toString() === dhService.serverUrl.toString();

    connectionOptions.push({
      type: 'connection',
      label: dhService.serverUrl.toString(),
      iconPath: new vscode.ThemeIcon(ICON_ID.connected),
      description: isActiveConnection
        ? `${editorLanguageId} (current)`
        : editorLanguageId,
      data: dhService,
    });
  }

  if (serverOptions.length === 0 && connectionOptions.length === 0) {
    throw new Error('No available servers to connect to.');
  }

  const options: (
    | SeparatorPickItem
    | ConnectionPickItem<'server', ServerState>
    | ConnectionPickItem<'connection', IDhService>
  )[] = [
    {
      label: 'Active Connections',
      kind: vscode.QuickPickItemKind.Separator,
    },
    ...connectionOptions,
    { label: 'Connect to Server', kind: vscode.QuickPickItemKind.Separator },
    ...serverOptions,
  ];

  const result = await vscode.window.showQuickPick(options, {
    title: 'Connect Editor',
  });

  if (result == null || !('type' in result)) {
    return null;
  }

  return result;
}

/**
 * Create a status bar item for connecting to DH server
 */
export function createConnectStatusBarItem(
  show: boolean
): vscode.StatusBarItem {
  const statusBarItem = vscode.window.createStatusBarItem(
    vscode.StatusBarAlignment.Right,
    2000
  );
  statusBarItem.command = SELECT_CONNECTION_COMMAND;
  const text = createConnectText('disconnected');
  statusBarItem.text = text;

  if (show) {
    statusBarItem.show();
  } else {
    statusBarItem.hide();
  }

  return statusBarItem;
}

/**
 * Create an option for the connection selection picker.
 * @param type The type of connection
 */
export function createConnectionOption(type: ConnectionType) {
  return (url: URL): ConnectionOption => {
    const label = `${type}: ${url.hostname}:${url.port}`;

    return { type, label, url };
  };
}

/**
 * Create display text for the connection status bar item.
 * @param status The connection status
 * @param option The connection option
 */
export function createConnectText(
  status: 'connecting' | 'connected' | 'disconnected',
  option?: ConnectionOption
): string {
  const icon = {
    connecting: `$(${ICON_ID.connecting})`,
    connected: `$(${ICON_ID.connected})`,
    disconnected: `$(${ICON_ID.disconnected})`,
  }[status];

  const label = {
    connecting: STATUS_BAR_CONNECTING_TEXT,
    connected: option?.label,
    disconnected: STATUS_BAR_DISCONNECTED_TEXT,
  }[status];

  const text = `${icon} ${label}`;

  return text;
}

/**
 * Format the connection label for display.
 * @param label The original label to format
 * @param isSelected Whether the connection is selected
 * @param consoleType The console type
 */
export function formatConnectionLabel(
  label: string,
  isSelected: boolean,
  consoleType?: ConsoleType
): string {
  const consoleTypeStr = consoleType ? ` (${consoleType})` : '';
  return isSelected
    ? `$(${ICON_ID.connected}) ${label}${consoleTypeStr}`
    : `$(${ICON_ID.blank}) ${label}${consoleTypeStr}`;
}

// Copied from @deephaven/console `ConsoleUtils`
export function formatTimestamp(date: Date): string | null {
  if (date == null || !(date instanceof Date)) {
    return null;
  }

  const hours = `${date.getHours()}`.padStart(2, '0');
  const minutes = `${date.getMinutes()}`.padStart(2, '0');
  const seconds = `${date.getSeconds()}`.padStart(2, '0');
  const milliseconds = `${date.getMilliseconds()}`.padStart(3, '0');

  return `${hours}:${minutes}:${seconds}.${milliseconds}`;
}

/**
 * Get a `TextEditor` containing the given uri. If there is one already open,
 * it will be returned. Otherwise, a new one will be opened. The returned editor
 * will become the active editor if it is not already.
 * @param uri
 */
export async function getEditorForUri(
  uri: vscode.Uri
): Promise<vscode.TextEditor> {
  if (
    uri.toString() === vscode.window.activeTextEditor?.document.uri.toString()
  ) {
    return vscode.window.activeTextEditor;
  }

  const viewColumn = vscode.window.visibleTextEditors.find(
    editor => editor.document.uri.toString() === uri.toString()
  )?.viewColumn;

  // If another panel such as the output panel is active, set the document
  // for the url to active first
  // https://stackoverflow.com/a/64808497/20489
  return vscode.window.showTextDocument(uri, { preview: false, viewColumn });
}

/**
 * Update given status bar item based on connection status
 * and optional `ConnectionOption`.
 * @param statusBarItem The status bar item to update
 * @param status The connection status
 * @param option The connection option
 */
export function updateConnectionStatusBarItem(
  statusBarItem: vscode.StatusBarItem | null | undefined,
  status: 'connecting' | 'connected' | 'disconnected',
  option?: ConnectionOption
): void {
  if (statusBarItem == null) {
    return;
  }

  const text = createConnectText(status, option);
  statusBarItem.text = text;
}
