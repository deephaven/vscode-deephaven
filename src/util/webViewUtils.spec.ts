import * as vscode from 'vscode';
import { beforeEach, describe, expect, it, vi } from 'vitest';
import {
  createGetPropertyResponseMsg,
  getWebViewContentRootUri,
  getWebViewHtml,
  registerWebViewThemeHandlers,
} from './webViewUtils';
import { VIEW_ID_PREFIX, type ViewID } from '../common';
import { mockT } from '../crossModule/testUtils';
import { uniqueId } from './idUtils';
import type { UniqueID } from '../types';
import {
  VSCODE_POST_MSG,
  type BaseThemeKey,
  type VscodeGetPropertyMsg,
  type VscodePropertyName,
} from '../crossModule';
import { getDHThemeKey } from './uiUtils';

vi.mock('vscode');
vi.mock('./idUtils');
vi.mock('./uiUtils');

beforeEach(() => {
  vi.clearAllMocks();
  vi.mocked(uniqueId).mockReturnValue('mock.uniqueId' as UniqueID);
});

const mockColorTheme = {} as vscode.ColorTheme;
const mockExtensionUri = {
  path: 'mock.extensionUri',
  toString: function () {
    return this.path;
  },
} as vscode.Uri;
const mockViewId = `${VIEW_ID_PREFIX}mockViewId` as ViewID;
const mockWebView = mockT<vscode.Webview>({
  cspSource: 'mock.cspSource',
  asWebviewUri: vi.fn(),
  onDidReceiveMessage: vi.fn(),
  postMessage: vi.fn(),
});
const mockServerUrl = new URL('https://mock.deephaven.io');
const mockWebViewView = mockT<vscode.WebviewView>({
  webview: mockWebView,
  onDidDispose: vi.fn(),
});

describe('createGetPropertyResponseMsg', () => {
  it('should return the correct message', () => {
    const id = 'mock.id';
    const name = 'mock.name' as VscodePropertyName;
    const value = 'mock.themeKey' as BaseThemeKey;

    vi.mocked(getDHThemeKey).mockReturnValue(value);

    const result = createGetPropertyResponseMsg(id, name);

    expect(result).toEqual({
      id,
      message: VSCODE_POST_MSG.getVscodePropertyResponse,
      payload: {
        name,
        value,
      },
    });
  });
});

describe('getWebViewContentRootUri', () => {
  it('should return the correct Uri', () => {
    const result = getWebViewContentRootUri(mockExtensionUri, mockViewId);

    expect(vscode.Uri.joinPath).toHaveBeenCalledWith(
      mockExtensionUri,
      'out',
      'webViews',
      mockViewId.replace(VIEW_ID_PREFIX, '')
    );

    const expected = vi.mocked(vscode.Uri.joinPath).mock.results[0].value;

    expect(result).toBe(expected);
  });
});

describe('getWebViewHtml', () => {
  beforeEach(() => {
    vi.mocked(mockWebView.asWebviewUri).mockImplementation(uri =>
      vscode.Uri.parse(`mockwebview://${uri}`)
    );
  });

  it('should return the correct HTML', () => {
    expect(
      getWebViewHtml({
        extensionUri: mockExtensionUri,
        webView: mockWebView,
        viewId: mockViewId,
        iframeUrl: new URL('https://example.com'),
        scriptFileName: 'script.js',
        stylesFileName: 'styles.css',
      })
    ).toMatchSnapshot();
  });
});

describe('registerWebViewThemeHandlers', () => {
  it('should register an `onDidChangeActiveColorTheme` event handler', () => {
    registerWebViewThemeHandlers(mockWebViewView, mockServerUrl);

    const handler = vi.mocked(vscode.window.onDidChangeActiveColorTheme).mock
      .calls[0][0];

    expect(handler).toBeInstanceOf(Function);

    handler(mockColorTheme);

    expect(mockWebView.postMessage).toHaveBeenCalledWith({
      id: uniqueId(),
      message: VSCODE_POST_MSG.requestSetTheme,
      payload: getDHThemeKey(),
      targetOrigin: mockServerUrl.origin,
    });
  });

  it('should register an `onDidReceiveMessage` event handler', () => {
    registerWebViewThemeHandlers(mockWebViewView, mockServerUrl);

    const handler = vi.mocked(mockWebView.onDidReceiveMessage).mock.calls[0][0];

    expect(handler).toBeInstanceOf(Function);

    const id = 'mock.id';
    const payload = 'baseThemeKey';

    const message: { data: VscodeGetPropertyMsg; origin: string } = {
      data: {
        id,
        message: VSCODE_POST_MSG.getVscodeProperty,
        payload,
      },
      origin: mockServerUrl.origin,
    };

    handler(message);

    expect(mockWebView.postMessage).toHaveBeenCalledWith(
      createGetPropertyResponseMsg(id, payload)
    );
  });

  it('should regster an `onDidDispose` event handler that cleans up subscriptions', () => {
    const colorChangeSubscription = { dispose: vi.fn() };
    const messageSubscription = { dispose: vi.fn() };

    vi.mocked(vscode.window.onDidChangeActiveColorTheme).mockReturnValue(
      colorChangeSubscription
    );

    vi.mocked(mockWebView.onDidReceiveMessage).mockReturnValue(
      messageSubscription
    );

    registerWebViewThemeHandlers(mockWebViewView, mockServerUrl);

    const disposeHandler = vi.mocked(mockWebViewView.onDidDispose).mock
      .calls[0][0];
    expect(disposeHandler).toBeInstanceOf(Function);

    disposeHandler();
    expect(colorChangeSubscription.dispose).toHaveBeenCalled();
    expect(messageSubscription.dispose).toHaveBeenCalled();
  });
});
