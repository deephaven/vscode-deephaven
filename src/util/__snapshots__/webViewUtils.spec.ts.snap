// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`getWebViewHtml > should return the correct HTML 1`] = `
"<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8">
  <meta http-equiv="Content-Security-Policy" content="default-src 'none'; style-src mock.cspSource; script-src 'nonce-mock.uniqueId'; frame-src https://example.com">
  <meta name="dh-iframe-url" content="https://example.com/">
  <title>DH WebView</title>
  <link rel="stylesheet" href="mockwebview://mock.extensionUri/out/webViews/mockViewId/styles.css">
</head>
<body>
  <script nonce="mock.uniqueId" src="mockwebview://mock.extensionUri/out/webViews/mockViewId/script.js"></script>
</body>
</html>"
`;
